onnxruntime/contrib_ops/cpu/bert/attention.cc

```diff
@@ -115,151 +115,175 @@ Status Attention<T>::Compute(OpKernelContext* context) const {
+      // C: QKV[qkv_index] (3xBxNxSxH)        (3.B.N.)S x H         S x H
+
+      math::GemmEx<float, concurrency::ThreadPool>(CblasNoTrans,                   // TransA = no
+                                                   CblasNoTrans,                   // TransB = no
+                                                   sequence_length,                // M      = S
+                                                   head_size,                      // N      = H
+                                                   hidden_size,                    // K      = NH
+                                                   1.0f,                           // alpha
+                                                   input_data + input_offset,      // A
+                                                   hidden_size,                    // lda    = NH
+                                                   weights_data + weights_offset,  // B
+                                                   3 * hidden_size,                // ldb    = 3NH
+                                                   1.0f,                           // beta
+                                                   qkv_dest + qkv_offset,          // C
+                                                   head_size,                      // ldc
+                                                   nullptr                         // use single-thread
+      );
+    });
   }
 
-  {
-    int offset_Q = 0;
-    int offset_Q_increment = sequence_length * head_size;
-    int offset_scratch = 0;
-    int offset_scratch_increment = sequence_length * sequence_length;
+  // STEP.2: scratch(B, N, S, S) = 1/sqrt(H) x Q(B, N, S, H) x K'(B, N, S, H -> B, N, H, S) + 1 x mask_index(B -> B, 1, 1, 1)
+  auto scratch_data = allocator->Alloc(batch_size * num_heads_ * sequence_length * sequence_length * element_size);
+  BufferUniquePtr scratch_buffer(scratch_data, BufferDeleter(allocator));
 
+  {
+    auto scratch_broadcast_data = allocator->Alloc(batch_size * sequence_length * element_size);
+    BufferUniquePtr scratch_broadcast_buffer(scratch_broadcast_data, BufferDeleter(allocator));
+    memset(scratch_broadcast_data, 0, batch_size * sequence_length * element_size);
+    T* p_scratch_broadcast_current_data = reinterpret_cast<T*>(scratch_broadcast_data);
     for (int b_i = 0; b_i < batch_size; b_i++) {
-      for (int n_i = 0; n_i < num_heads_; n_i++) {
-        math::Gemm<T>(
-            CblasNoTrans,
-            CblasTrans,
-            sequence_length,
-            sequence_length,
-            head_size,
-            1.0f / sqrt(static_cast<float>(head_size)),
-            Q + offset_Q,
-            K + offset_Q,
-            1.0,
-            reinterpret_cast<T*>(scratch_data) + offset_scratch,
-            tp);
-        offset_Q += offset_Q_increment;
-        offset_scratch += offset_scratch_increment;
+      int mask = mask_index->template Data<int>()[b_i];
```

Comment:
int32_t

Reply:
mask_index can be used in softmax to save some calculation. See cuda kernel usage. Just a hint for further improvement in the future.

Reply:
I added comments to make this as a TODO item